files/routes/chat.py

```diff
@@ -1,117 +1,271 @@
+import functools
 import time
-from files.helpers.config.environment import SITE, SITE_FULL
-from files.helpers.wrappers import auth_required
-from files.helpers.sanitize import sanitize
+import uuid
+from typing import Any, Final
+
+from flask_socketio import SocketIO, emit, disconnect
+
+from files.__main__ import app, cache, limiter
+from files.helpers.alerts import *
 from files.helpers.config.const import *
-from datetime import datetime
-from flask_socketio import SocketIO, emit
-from files.__main__ import app, limiter, cache
-from flask import render_template, make_response, send_from_directory
-import sys
-import atexit
-
-if SITE == 'localhost':
-	socketio = SocketIO(app, async_mode='gevent', cors_allowed_origins=[SITE_FULL], logger=True, engineio_logger=True, debug=True)
+from files.helpers.config.environment import *
+from files.helpers.config.regex import *
+from files.helpers.sanitize import sanitize
+from files.helpers.wrappers import get_logged_in_user, is_not_permabanned, admin_level_required
+
+def chat_is_allowed(perm_level: int=0):
+	def wrapper_maker(func):
+		@functools.wraps(func)
+		def wrapper(*args: Any, **kwargs: Any) -> bool | None:
+			v = get_logged_in_user()
+			if not v:
+				return abort(403)
```

Comment:
abort() doesn't return, wut.

Reply:
lol whoops

I had a bunch of trouble with getting this working and some weird stuff happened in the process. Eventually I realized the chat code wasn't auto-updating :V